File: backend/src/services/langgraph/state.ts
import { BaseMessage } from '@langchain/core/messages';

// Use type-only imports to avoid triggering TypeORM initialization
import type { Persona } from '@/entities/Persona';
import type { Simulation } from '@/entities/Simulation';
import type { SimulationSession } from '@/entities/SimulationSession';

/**
 * Proactive message trigger types
 */
export type ProactiveTrigger = 'followup' | 'inactivity' | 'backchannel' | 'start';

/**
 * Goal progress status for tracking conversation objectives
 */
export interface GoalProgressItem {
  goalNumber: number;
  isOptional: boolean;
  title: string;
  status: 'not_started' | 'in_progress' | 'achieved';
  confidence: number;
  evidence?: Array<{
    messageId: string;
    role: 'user' | 'ai';
    label: string;
    score: number;
  }>;
  startedAt?: string;
  achievedAt?: string;
}

/**
 * Main state schema for the conversation graph
 * This state is passed between nodes and persisted via checkpoints
 */
export interface ConversationGraphState {
  // Session identification
  sessionId: string;
  userId: string;
  
  // Conversation history in LangChain format
  messages: BaseMessage[];
  
  // Persona and simulation context
  persona: Persona;
  simulation: Simulation;
  
  // Goal tracking
  goalProgress: GoalProgressItem[];
  
  // Turn management
  turn: 'user' | 'ai';
  
  // Last messages for context
  lastUserMessage?: string;
  lastAiMessage?: string;
  
  // Proactive message handling
  proactiveTrigger?: ProactiveTrigger;
  shouldSendProactive: boolean;
  proactiveCount: number;
  maxProactiveMessages?: number;
  
  // RAG context
  ragContext?: string;
  
  // Evaluation flags
  needsEvaluation: boolean;
  evaluationComplete: boolean;
  
  // Analysis results (from Transformers microservice)
  lastEmotionAnalysis?: {
    emotion: string;
    confidence: number;
  };
  lastSentimentAnalysis?: {
    sentiment: 'positive' | 'neutral' | 'negative';
    confidence: number;
  };
  lastQualityScores?: {
    overall?: number;
    coherence?: number;
    relevance?: number;
    completeness?: number;
    personaAlignment?: number;
  };
  
  // Metadata for additional context
  metadata: {
    startedAt?: Date;
    sessionDuration?: number;
    messageCount?: number;
    aiInitiated?: boolean;
    lastAiMessageAt?: Date;
    lastUserMessageAt?: Date;
    inactivityNudgeAt?: Date;
    inactivityNudgeCount?: number;
    conversationStyle?: any;
    [key: string]: any;
  };
  
  // Error handling
  lastError?: string;
  retryCount?: number;
  
  // Checkpoint management
  checkpointId?: string;
  checkpointVersion?: number;
}

/**
 * Input schema for starting a conversation or sending a message
 */
export interface ConversationInput {
  sessionId: string;
  userId: string;
  userMessage?: string;
  proactiveTrigger?: ProactiveTrigger;
  metadata?: Record<string, any>;
}

/**
 * Output schema for graph execution results
 */
export interface ConversationOutput {
  sessionId: string;
  messages: BaseMessage[];
  lastAiMessage?: string;
  goalProgress: GoalProgressItem[];
  turn: 'user' | 'ai';
  metadata: {
    confidence?: number;
    processingTime?: number;
    emotionalTone?: string;
    sentiment?: 'positive' | 'neutral' | 'negative';
    [key: string]: any;
  };
}

/**
 * Helper to convert SimulationSession to initial graph state
 */
export function sessionToGraphState(
  session: SimulationSession,
  persona: Persona,
  simulation: Simulation,
  messages: BaseMessage[] = [],
): Partial<ConversationGraphState> {
  return {
    sessionId: session.id,
    userId: (session.user as any)?.id || '',
    messages,
    persona,
    simulation,
    goalProgress: session.goalProgress as GoalProgressItem[] || [],
    turn: session.turn || 'ai',
    shouldSendProactive: false,
    proactiveCount: 0,
    needsEvaluation: false,
    evaluationComplete: false,
    metadata: {
      startedAt: session.startedAt,
      sessionDuration: session.startedAt ? Date.now() - session.startedAt.getTime() : 0,
      messageCount: session.messageCount || 0,
      aiInitiated: session.aiInitiated || false,
      lastAiMessageAt: session.lastAiMessageAt,
      lastUserMessageAt: session.lastUserMessageAt,
      inactivityNudgeAt: session.inactivityNudgeAt,
      inactivityNudgeCount: session.inactivityNudgeCount || 0,
      conversationStyle: persona.conversationStyle,
    },
  };
}

/**
 * Helper to extract key metrics from state for logging/monitoring
 */
export function extractStateMetrics(state: ConversationGraphState) {
  return {
    sessionId: state.sessionId,
    messageCount: state.messages.length,
    turn: state.turn,
    goalsAchieved: state.goalProgress.filter(g => g.status === 'achieved').length,
    totalGoals: state.goalProgress.length,
    proactiveCount: state.proactiveCount,
    evaluationComplete: state.evaluationComplete,
    hasError: !!state.lastError,
    retryCount: state.retryCount || 0,
  };
}

File: agent/src/careersim_agent/graph/state.py
"""State schema for the conversation graph."""

from typing import TypedDict, Literal, Optional, Annotated, Any
from langchain_core.messages import BaseMessage
from langgraph.graph.message import add_messages


class EvidenceItem(TypedDict, total=False):
    """Evidence for goal progress."""
    messageIndex: int
    role: Literal["user", "ai"]
    label: str
    score: float


class GoalProgressItem(TypedDict, total=False):
    """Progress tracking for a conversation goal."""
    goalNumber: int
    isOptional: bool
    title: str
    status: Literal["not_started", "in_progress", "achieved"]
    confidence: float
    evidence: list[EvidenceItem]
    startedAt: Optional[str]
    achievedAt: Optional[str]


class AnalysisResult(TypedDict, total=False):
    """NLP analysis result."""
    label: str  # emotion or sentiment label
    confidence: float
    source: str


class NodeTraceEntry(TypedDict):
    """Entry in the node execution trace."""
    node: str
    timestamp: str
    duration_ms: float
    input_summary: str
    output_summary: str


class ConversationState(TypedDict, total=False):
    """Main state schema for the conversation graph.
    
    This state is passed between nodes and maintains the conversation context.
    """
    # Session identification
    session_id: str
    
    # Conversation history (uses LangGraph's message reducer)
    messages: Annotated[list[BaseMessage], add_messages]
    
    # Persona and simulation context (loaded from JSON)
    persona: dict[str, Any]
    simulation: dict[str, Any]
    
    # Goal tracking
    goal_progress: list[GoalProgressItem]
    
    # Turn management: whose turn it is NEXT
    # 'user' = AI just spoke, waiting for user
    # 'ai' = User just spoke, AI should respond
    turn: Literal["user", "ai"]
    
    # Last messages for quick access
    last_user_message: Optional[str]
    last_ai_message: Optional[str]
    
    # Input field for new user messages
    user_message: Optional[str]
    
    # Proactive message handling
    proactive_trigger: Optional[Literal["followup", "inactivity", "start"]]
    should_send_proactive: bool
    proactive_count: int
    max_proactive_messages: int
    
    # Analysis results for user messages
    last_user_emotion: Optional[AnalysisResult]
    last_user_sentiment: Optional[AnalysisResult]
    
    # Analysis results for AI messages
    last_ai_emotion: Optional[AnalysisResult]
    last_ai_sentiment: Optional[AnalysisResult]
    
    # Evaluation flags
    needs_evaluation: bool
    evaluation_complete: bool
    
    # Metadata
    message_count: int
    started_at: Optional[str]
    
    # RAG retrieved context
    retrieved_context: Optional[str]
    
    # Error handling
    last_error: Optional[str]
    
    # Debug/tracing
    node_trace: list[NodeTraceEntry]


def create_initial_state(
    session_id: str,
    simulation: dict[str, Any],
    persona: dict[str, Any],
) -> ConversationState:
    """Create initial state for a new conversation.
    
    Args:
        session_id: Unique session identifier
        simulation: Simulation configuration
        persona: Persona configuration
        
    Returns:
        Initial conversation state
    """
    from datetime import datetime
    
    # Initialize goal progress from simulation goals
    goals = simulation.get("conversationGoals", [])
    goal_progress = [
        GoalProgressItem(
            goalNumber=goal["goalNumber"],
            isOptional=goal.get("isOptional", False),
            title=goal["title"],
            status="not_started",
            confidence=0.0,
            evidence=[],
        )
        for goal in goals
    ]
    
    return ConversationState(
        session_id=session_id,
        messages=[],
        persona=persona,
        simulation=simulation,
        goal_progress=goal_progress,
        turn="ai",  # AI starts (if startsConversation is true)
        last_user_message=None,
        last_ai_message=None,
        user_message=None,
        proactive_trigger="start" if persona.get("conversationStyle", {}).get("startsConversation") else None,
        should_send_proactive=False,
        proactive_count=0,
        max_proactive_messages=2,
        last_user_emotion=None,
        last_user_sentiment=None,
        last_ai_emotion=None,
        last_ai_sentiment=None,
        retrieved_context=None,
        needs_evaluation=False,
        evaluation_complete=False,
        message_count=0,
        started_at=datetime.now().isoformat(),
        last_error=None,
        node_trace=[],
    )


def get_current_goal(state: ConversationState) -> Optional[dict]:
    """Get the current goal being worked on.
    
    Returns the first unachieved required goal, or first unachieved optional goal
    if all required goals are achieved.
    """
    goals = state.get("simulation", {}).get("conversationGoals", [])
    progress = {p["goalNumber"]: p for p in state.get("goal_progress", [])}
    
    # Sort goals by number
    sorted_goals = sorted(goals, key=lambda g: g["goalNumber"])
    
    # Find first unachieved required goal
    for goal in sorted_goals:
        if goal.get("isOptional"):
            continue
        p = progress.get(goal["goalNumber"], {})
        if p.get("status") != "achieved":
            return goal
    
    # All required done, find first unachieved optional
    for goal in sorted_goals:
        if not goal.get("isOptional"):
            continue
        p = progress.get(goal["goalNumber"], {})
        if p.get("status") != "achieved":
            return goal
    
    return None  # All goals achieved
